File: C/0009.c
#include <stdbool.h>
#include <limits.h>

bool isPalindrome(int x) {
    // Negative numbers are not palindromes
    if (x < 0) return false;

    // Variables for reversed number and original number
    int original = x;
    int reversed = 0;

    // Reverse the number
    while (x > 0) {
        int digit = x % 10;
        
        // Check for potential overflow before multiplying and adding
        if (reversed > (INT_MAX - digit) / 10) {
            return false; // Overflow would occur
        }

        reversed = reversed * 10 + digit;
        x /= 10;
    }

    // Check if the reversed number is equal to the original number
    return original == reversed;
}

File: Dart/0009.dart
class Solution {
  bool isPalindrome(int x) {
    // Negative numbers and numbers ending with zero (but not zero itself) cannot be palindromes.
    if (x < 0 || (x % 10 == 0 && x != 0)) {
      return false;
    }
    
    int original = x;
    int reversed = 0;
    
    while (x > 0) {
      int digit = x % 10; // Get the last digit
      reversed = reversed * 10 + digit; // Append digit to the reversed number
      x ~/= 10; // Remove the last digit from x
    }
    
    return original == reversed; // Check if the original number is the same as the reversed number
  }
}
